pyteal/ast/scratch.py

```diff
@@ -9,7 +9,58 @@
     from ..compiler import CompileOptions
 
 
-class ScratchSlot:
+class Slot:
+    """Abstract Slot class"""
+
+    def slotIdFromStack(self) -> bool:
+        pass
+
+    def store(self, value: Expr = None) -> Expr:
+        """Get an expression to store a value in this slot.
+
+        Args:
+            value (optional): The value to store in this slot. If not included, the last value on
+            the stack will be stored. NOTE: storing the last value on the stack breaks the typical
+            semantics of PyTeal, only use if you know what you're doing.
+        """
+        if value is not None:
+            return ScratchStore(self, value)
+        return ScratchStackStore(self)
```

Comment:
Gotta figure out the impact of this `ScratchStackStore` in the case of no value provided... not sure it works out of the box with `DynamicSlot`